client/my-sites/site-settings/site-icon-setting/index.jsx

```diff
@@ -76,22 +78,36 @@ class SiteIconSetting extends Component {
 			return;
 		}
 
+		// Upload media using a manually generated ID so that we can continue
+		// to reference it within this function
 		const transientMediaId = uniqueId( 'site-icon' );
-
 		MediaActions.add( siteId, {
 			ID: transientMediaId,
 			fileName: `cropped-${ selectedItem.file }`,
 			fileContents: blob
 		} );
 
+		// Avoid calling to component instance within progress check closure
+		// below since component may have been unmounted in time since upload
+		const {
+			saveSiteSettings: dispatchSaveSiteSettings,
+			setSiteIcon: dispatchSetSiteIcon
+		} = this.props;
```

Comment:
>don't you think the behaviour here should be moved to a thunk, It has nothing that relates it to the component, so maybe a thunk like `uploadSiteIcon( filename, blob )` that could call your `uploadFile` from #9706 ?

Yeah, I think ideally that'd be the case. The big issue is interoperability with the previous Flux implementation. With #9706 we still don't have validation, and we still don't have the media items showing up in the media library listing (so after the image is uploaded, it won't show until refreshing the page).

I might see if I can at least consolidate or wrap some of the component dispatching logic here so it becomes less aware of the specific actions its effecting.